File: android/app/src/main/java/com/openavc/panel/discovery/CertFetcher.kt
package com.openavc.panel.discovery

import android.content.Context
import android.util.Log
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.net.HttpURLConnection
import java.net.URL
import java.security.cert.X509Certificate
import javax.net.ssl.HostnameVerifier
import javax.net.ssl.HttpsURLConnection
import javax.net.ssl.SSLContext
import javax.net.ssl.X509TrustManager

/**
 * First-contact certificate retrieval for OpenAVC HTTPS servers.
 *
 * Auto-generate TLS:
 *   GET https://host:port/api/certificate returns the server's CA in PEM.
 *   We pin that — the leaf rotates freely without re-pair.
 *
 * Provided TLS (user uploaded their own cert):
 *   /api/certificate returns 404. We open a trust-all handshake against
 *   /api/health and capture the leaf cert presented by the server, then pin it.
 *   A future leaf rotation requires re-pair (acceptable for the BYO-cert path).
 *
 * Both flows are TOFU on first contact. Mitigation: admin sheet shows the
 * pinned fingerprint so a paranoid integrator can compare against the value
 * shown in the Programmer IDE.
 */
class CertFetcher(context: Context) {

    private val trustStore = CertTrustStore(context)

    suspend fun fetchAndPin(
        host: String,
        port: Int,
        instanceId: String? = null,
    ): X509Certificate? = withContext(Dispatchers.IO) {
        val hostPort = CertTrustStore.hostPortKey(host, port)
        val auto = fetchAutoCa(host, port)
        if (auto != null) {
            trustStore.pin(instanceId, hostPort, auto.pem.toByteArray(Charsets.UTF_8))
            return@withContext auto.cert
        }
        val leaf = fetchProvidedLeaf(host, port)
        if (leaf != null) {
            trustStore.pin(instanceId, hostPort, pemEncode(leaf).toByteArray(Charsets.UTF_8))
            return@withContext leaf
        }
        null
    }

    private fun fetchAutoCa(host: String, port: Int): FetchedPem? {
        val url = URL("https://$host:$port/api/certificate")
        val conn = openTrustAll(url) ?: return null
        try {
            conn.connectTimeout = TIMEOUT_MS
            conn.readTimeout = TIMEOUT_MS
            conn.requestMethod = "GET"
            val code = conn.responseCode
            if (code == 404) return null
            if (code !in 200..299) {
                Log.d(TAG, "auto-ca probe got $code from $url")
                return null
            }
            val pem = conn.inputStream.bufferedReader().use { it.readText() }
            val cert = CertTrustStore.parsePem(pem)
            return FetchedPem(cert, pem)
        } catch (e: Exception) {
            Log.d(TAG, "auto-ca fetch failed: ${e.message}")
            return null
        } finally {
            conn.disconnect()
        }
    }

    private fun fetchProvidedLeaf(host: String, port: Int): X509Certificate? {
        val url = URL("https://$host:$port/api/health")
        val conn = openTrustAll(url) ?: return null
        try {
            conn.connectTimeout = TIMEOUT_MS
            conn.readTimeout = TIMEOUT_MS
            conn.requestMethod = "GET"
            conn.connect()
            val peer = conn.serverCertificates.firstOrNull() as? X509Certificate
            if (peer == null) {
                Log.d(TAG, "no peer cert from $url")
                return null
            }
            // Drain the body so the connection can be pooled cleanly.
            runCatching { conn.inputStream.use { it.readBytes() } }
            return peer
        } catch (e: Exception) {
            Log.d(TAG, "provided-leaf fetch failed: ${e.message}")
            return null
        } finally {
            conn.disconnect()
        }
    }

    private fun openTrustAll(url: URL): HttpsURLConnection? {
        val conn = url.openConnection() as? HttpsURLConnection ?: return null
        val ctx = SSLContext.getInstance("TLS")
        ctx.init(null, arrayOf<X509TrustManager>(TRUST_ALL), java.security.SecureRandom())
        conn.sslSocketFactory = ctx.socketFactory
        conn.hostnameVerifier = HostnameVerifier { _, _ -> true }
        return conn
    }

    private data class FetchedPem(val cert: X509Certificate, val pem: String)

    companion object {
        private const val TAG = "CertFetcher"
        private const val TIMEOUT_MS = 3_000

        private val TRUST_ALL = object : X509TrustManager {
            override fun checkClientTrusted(chain: Array<out X509Certificate>?, authType: String?) {}
            override fun checkServerTrusted(chain: Array<out X509Certificate>?, authType: String?) {}
            override fun getAcceptedIssuers(): Array<X509Certificate> = emptyArray()
        }

        fun pemEncode(cert: X509Certificate): String {
            val b64 = java.util.Base64.getEncoder().encodeToString(cert.encoded)
            val wrapped = b64.chunked(64).joinToString("\n")
            return "-----BEGIN CERTIFICATE-----\n$wrapped\n-----END CERTIFICATE-----\n"
        }
    }
}

File: ios/OpenAVCPanel/Discovery/CertFetcher.swift
import Foundation
import Security

/// First-contact certificate retrieval for OpenAVC HTTPS servers.
///
/// Auto-generate TLS:
///   `GET https://host:port/api/certificate` returns the server's CA in PEM.
///   We pin that, so the leaf can rotate without a re-pair.
///
/// Provided TLS (the integrator uploaded their own certificate):
///   `/api/certificate` answers 404. We open a trust-all handshake against
///   `/api/health`, capture the leaf the server presents, and pin that instead.
///   A later leaf rotation needs a re-pair, which is acceptable for BYO-cert.
///
/// Both are trust-on-first-use. The mitigation is the same as Android's: the
/// admin sheet shows the pinned fingerprint to compare against the IDE.
struct CertFetcher {
    private let trustStore: CertTrustStore
    private static let timeout: TimeInterval = 3

    init(trustStore: CertTrustStore = CertTrustStore()) {
        self.trustStore = trustStore
    }

    func fetchAndPin(host: String, port: Int, instanceId: String? = nil) async -> SecCertificate? {
        let hostPort = CertTrustStore.hostPortKey(host: host, port: port)

        if let auto = await fetchAutoCa(host: host, port: port) {
            trustStore.pin(instanceId: instanceId, hostPort: hostPort, pem: auto.pem)
            return auto.certificate
        }
        if let leaf = await fetchProvidedLeaf(host: host, port: port) {
            trustStore.pin(
                instanceId: instanceId,
                hostPort: hostPort,
                pem: CertTrustStore.pemEncode(leaf)
            )
            return leaf
        }
        return nil
    }

    private struct FetchedPem {
        let pem: String
        let certificate: SecCertificate
    }

    private func fetchAutoCa(host: String, port: Int) async -> FetchedPem? {
        guard let url = URL(string: "https://\(host):\(port)/api/certificate") else { return nil }
        let delegate = TrustAllSessionDelegate()
        let session = Self.session(delegate: delegate)
        defer { session.finishTasksAndInvalidate() }

        do {
            let (data, response) = try await session.data(for: Self.request(url))
            guard let http = response as? HTTPURLResponse, (200..<300).contains(http.statusCode) else {
                return nil
            }
            let pem = String(decoding: data, as: UTF8.self)
            guard let certificate = CertTrustStore.parsePem(pem) else { return nil }
            return FetchedPem(pem: pem, certificate: certificate)
        } catch {
            return nil
        }
    }

    private func fetchProvidedLeaf(host: String, port: Int) async -> SecCertificate? {
        guard let url = URL(string: "https://\(host):\(port)/api/health") else { return nil }
        let delegate = TrustAllSessionDelegate()
        let session = Self.session(delegate: delegate)
        defer { session.finishTasksAndInvalidate() }

        // The response body is irrelevant; the handshake is the point. The
        // delegate captures the leaf on its way past.
        _ = try? await session.data(for: Self.request(url))
        return delegate.capturedLeaf
    }

    private static func request(_ url: URL) -> URLRequest {
        var request = URLRequest(url: url)
        request.timeoutInterval = timeout
        request.httpMethod = "GET"
        return request
    }

    private static func session(delegate: URLSessionDelegate) -> URLSession {
        let config = URLSessionConfiguration.ephemeral
        config.timeoutIntervalForRequest = timeout
        config.timeoutIntervalForResource = timeout
        return URLSession(configuration: config, delegate: delegate, delegateQueue: nil)
    }
}
